Cubical/Algebra/CommRing/FGIdeal.agda

```diff
@@ -100,5 +107,146 @@ syntax genIdeal R V = ⟨ V ⟩[ R ]
 
 
 FGIdealIn : (R : CommRing ℓ) → Type (ℓ-suc ℓ)
-FGIdealIn R = Σ[ I ∈ CommIdeal R ] ∃[ n ∈ ℕ ] ∃[ V ∈ FinVec (fst R) n ] I ≡ ⟨ V ⟩[ R ]
-
+FGIdealIn R = Σ[ I ∈ CommIdeal.CommIdeal R ] ∃[ n ∈ ℕ ] ∃[ V ∈ FinVec (fst R) n ] I ≡ ⟨ V ⟩[ R ]
+
+-- The lattice laws
+module _ (R' : CommRing ℓ) where
+ open CommRingStr (snd R')
+ open RingTheory (CommRing→Ring R')
+ open CommIdeal R'
+ open Sum (CommRing→Ring R')
+ open KroneckerDelta (CommRing→Ring R')
+ private
+  R = fst R'
+  ⟨_⟩ : {n : ℕ} → FinVec R n → CommIdeal
+  ⟨ V ⟩ = ⟨ V ⟩[ R' ]
+
+ inclOfFGIdeal : {n : ℕ} (V : FinVec R n) (I : CommIdeal)
+     → (∀ i → V i ∈ I .fst) → ⟨ V ⟩ .fst ⊆ I .fst
+ inclOfFGIdeal V I ∀i→Vi∈I x = elim (λ _ → I .fst x .snd) inclOfFGIdealΣ
+  where
+  inclOfFGIdealΣ : Σ[ α ∈ FinVec R _ ] x ≡ linearCombination R' α V → x ∈ I .fst
+  inclOfFGIdealΣ (α , x≡α·V) = subst-∈ (I .fst) (sym x≡α·V) (∑Closed I (λ i → α i · V i)
+                             λ i → ·Closed (I .snd) _ (∀i→Vi∈I i))
+
+ indInIdeal : ∀ {n : ℕ} (U : FinVec R n) (i : Fin n) → U i ∈ ⟨ U ⟩ .fst
+ indInIdeal U i = ∣ (δ i) , sym (∑Mul1r _ U i) ∣
+
+ sucIncl : ∀ {n : ℕ} (U : FinVec R (ℕsuc n)) → ⟨ U ∘ suc ⟩ .fst ⊆ ⟨ U ⟩ .fst
+ sucIncl U x = map λ (α , x≡∑αUsuc) → (λ { zero → 0r ; (suc i) → α i }) , x≡∑αUsuc ∙ path _ _
+  where
+  path : ∀ s u₀ → s ≡ 0r · u₀ + s
+  path = solve R'
+
+ emptyFGIdeal : ∀ (V : FinVec R 0) → ⟨ V ⟩ ≡ 0Ideal
+ emptyFGIdeal V = CommIdeal≡Char incl1 incl2
+  where
+  incl1 : ⟨ V ⟩ .fst ⊆ 0Ideal .fst
+  incl1 x = rec (is-set _ _) λ (_ , p) → p
+
+  incl2 : 0Ideal .fst ⊆ ⟨ V ⟩ .fst
+  incl2 x x≡0 = ∣ (λ ()) , x≡0 ∣
+
+ 0FGIdeal : {n : ℕ} → ⟨ replicateFinVec n 0r ⟩ ≡ 0Ideal
+ 0FGIdeal = CommIdeal≡Char incl1 incl2
+  where
+  incl1 : ⟨ replicateFinVec _ 0r ⟩ .fst ⊆ 0Ideal .fst
+  incl1 x = elim (λ _ → is-set _ _)
+          λ (α , x≡∑α0) → subst-∈ (0Ideal .fst) (sym x≡∑α0) (∑Closed 0Ideal (λ i → α i · 0r)
+          λ i → subst-∈ (0Ideal .fst) (sym (0RightAnnihilates _)) refl)
+
+  incl2 : 0Ideal .fst ⊆ ⟨ replicateFinVec _ 0r ⟩ .fst
+  incl2 x x≡0 = subst-∈ (⟨ replicateFinVec _ 0r ⟩ .fst) (sym x≡0)
+                         (⟨ replicateFinVec _ 0r ⟩ .snd .contains0)
+
+
+ -- better syntax for ∑ λ i → ... ???
```

Comment:
Remove this comment?

Reply:
Yeah, do you have an idea for nice `bigOp` syntax?

Reply:
Not really... I think it's pretty good as is :-)